src/cluster_slot.c

```diff
@@ -0,0 +1,280 @@
+    clusterNode *n = server.cluster->myself;
+    return (server.cluster->slots[slot] == n) || (nodeIsSlave(n) && (server.cluster->slots[slot] == n->slaveof));
+}
+
+static void markAssignedSlots(unsigned char *slots) {
+    for (int slot = 0; slot < CLUSTER_SLOTS; slot++) {
+        if (doesSlotBelongToCurrentShard(slot)) slots[slot]++;
+    }
+}
+
+static int countAssignedSlotsFromSlotsArray(unsigned char *slots) {
+    int count = 0;
+    for (int slot = 0; slot < CLUSTER_SLOTS; slot++) {
+        if (slots[slot]) count++;
+    }
+    return count;
+}
+
+static int countAssignedSlotsFromEntireShard() {
+    int count = 0;
+    for (int slot = 0; slot < CLUSTER_SLOTS; slot++) {
+        if (doesSlotBelongToCurrentShard(slot)) count++;
+    }
+    return count;
+}
+
+static int checkSlotAssignment(client *c, unsigned char *slots, int start_slot, int end_slot) {
+    for (int slot = start_slot; slot <= end_slot; slot++) {
+        if (doesSlotBelongToCurrentShard(slot)) {
+            if (slots[slot]++ == 1) {
+                addReplyErrorFormat(c,"Slot %d specified multiple times",(int)slot);
+                return C_ERR;
+            }
+        }
+    }
+    return C_OK;
+}
+
+static uint64_t getSingleSlotStat(int slot, int order_by) {
+    serverAssert(order_by != ORDER_BY_INVALID);
+    uint64_t singleSlotStat = 0;
+    if (order_by == ORDER_BY_KEY_COUNT) {
+        singleSlotStat = (*server.db->slots_to_keys).by_slot[slot].count;
+    }
+    return singleSlotStat;
+}
+
+static int slotStatEntryAscCmp(const void *a, const void *b) {
+    sortedSlotStatEntry entry_a = *((sortedSlotStatEntry *) a);
+    sortedSlotStatEntry entry_b = *((sortedSlotStatEntry *) b);
+    return entry_a.stat - entry_b.stat;
+}
+
+static int slotStatEntryDescCmp(const void *a, const void *b) {
+    sortedSlotStatEntry entry_a = *((sortedSlotStatEntry *) a);
+    sortedSlotStatEntry entry_b = *((sortedSlotStatEntry *) b);
+    return entry_b.stat - entry_a.stat;
+}
+
+static void sortSlotStats(sortedSlotStatEntry sorted[], int order_by, int desc) {
+    for (int slot = 0; slot < CLUSTER_SLOTS; slot++) {
+        if (doesSlotBelongToCurrentShard(slot)) {
+            sorted[slot].slot = slot;
+            sorted[slot].stat = getSingleSlotStat(slot, order_by);
+        } else {
+            /* Even if the slot does not belong to the current shard,
+             * we should fill the entry with default values so that qsort() does not segfault.
+             * These entries will be filtered and ignored upon addReplySortedSlotStats(). */
+            sorted[slot].slot = DEFAULT_SLOT;
+            sorted[slot].stat = DEFAULT_STAT;
+        }
+    }
+    qsort(sorted, CLUSTER_SLOTS, sizeof(sortedSlotStatEntry), (desc) ? slotStatEntryDescCmp : slotStatEntryAscCmp);
+}
+
+static void addReplySingleSlotStat(client *c, int slot) {
+    addReplyLongLong(c, slot);
+    addReplyMapLen(c, 1);
+    addReplyBulkCString(c, "key_count");
```

Comment:
```suggestion
    addReplyBulkCString(c, "key-count");
```

A question to pin for later. We sometimes use `-` and sometimes use `_`, we should probably pick a standard.

Reply:
i think most "proper" interfaces use `-` (e.g. MEMORY STATS, XINFO, ACL LOG).
the exception is INFO command that's often about internal variables.

Reply:
The one other exception I think was key specs in `INFO DOCS`.